workbench-core/authorization/src/dynamicAuthorization/dynamicAuthorizationService.test.ts

```diff
@@ -0,0 +1,347 @@
+      });
+
+      expect(mockDynamicPermissionsPlugin.getIdentityPermissionsBySubject).toHaveBeenNthCalledWith(1, {
+        subjectType: mockDynamicOperation.subjectType,
+        subjectId: mockDynamicOperation.subjectId,
+        action: mockDynamicOperation.action,
+        identities: [
+          { identityType: 'USER', identityId: mockAuthenticatedUser.id },
+          { identityType: 'GROUP', identityId: 'groupId1' },
+          { identityType: 'GROUP', identityId: 'groupId0' }
+        ]
+      });
+      expect(mockDynamicPermissionsPlugin.getIdentityPermissionsBySubject).toHaveBeenNthCalledWith(2, {
+        subjectType: mockDynamicOperation.subjectType,
+        subjectId: '*',
+        action: mockDynamicOperation.action,
+        identities: [
+          { identityType: 'USER', identityId: mockAuthenticatedUser.id },
+          { identityType: 'GROUP', identityId: 'groupId1' },
+          { identityType: 'GROUP', identityId: 'groupId0' }
+        ]
+      });
+
+      expect(mockAuthorizationPlugin.isAuthorizedOnDynamicOperations).toBeCalledWith(
+        [...sampleGroupPermissions, ...sampleGroupWildcardPermissions],
+        [mockDynamicOperation]
+      );
+    });
+
+    test('Check user with invalid permissions for a dynamic operation', async () => {
+      jest
+        .spyOn(mockAuthorizationPlugin, 'isAuthorizedOnDynamicOperations')
+        .mockRejectedValue(new ForbiddenError('User cannot access'));
+      try {
+        await dynamicAuthorizationService.isAuthorizedOnSubject(mockAuthenticatedUser, {
+          dynamicOperation: mockDynamicOperation
+        });
+        expect.hasAssertions();
+      } catch (err) {
+        expect(err).toBeInstanceOf(ForbiddenError);
+        expect(err.message).toBe('User cannot access');
+      }
+    });
+
+    test('No values returned for permissions for a dynamic operation', async () => {
+      jest.resetAllMocks();
+      jest.spyOn(mockDynamicPermissionsPlugin, 'getUserGroups').mockResolvedValueOnce({
+        groupIds: []
+      });
+
+      jest.spyOn(mockDynamicPermissionsPlugin, 'getIdentityPermissionsBySubject').mockResolvedValueOnce({
+        identityPermissions: []
+      });
+
+      jest.spyOn(mockDynamicPermissionsPlugin, 'getIdentityPermissionsBySubject').mockResolvedValueOnce({
+        identityPermissions: []
+      });
+      jest
+        .spyOn(mockAuthorizationPlugin, 'isAuthorizedOnDynamicOperations')
+        .mockRejectedValue(new ForbiddenError('User cannot access'));
+      try {
+        await dynamicAuthorizationService.isAuthorizedOnSubject(mockAuthenticatedUser, {
+          dynamicOperation: mockDynamicOperation
+        });
+        expect.hasAssertions();
+      } catch (err) {
+        expect(err).toBeInstanceOf(ForbiddenError);
+        expect(err.message).toBe('User cannot access');
+      }
+    });
+  });
+
+  describe('createGroup', () => {
+    test('Check for dynamic permissions plugin was called', async () => {
```

Comment:
I'm not sure we want to be testing implementation instead of outcomes. If the implementation changes, this test will fail, so the only thing this test really tells us is "these methods were called".

Reply:
I agree with this